File: macOS Source/TotallyHuman/Utilities/Localization.swift
// Localization.swift — Zweisprachige Oberfläche (Deutsch / Englisch)
// Totally Human — KI-Musik-Detektor (native macOS)
//
// Zentrale, sehr einfache Lokalisierung: Alle sichtbaren Texte der Oberfläche
// liegen hier als Schlüssel -> (Deutsch, Englisch) vor. Die aktuell gewählte
// Sprache steckt in `AppState.selectedLanguage` ("de" oder "en").
//
// Verwendung in den Views:   appState.t("schlüssel")
// Da `selectedLanguage` ein @Published-Wert ist, aktualisiert SwiftUI die
// Oberfläche automatisch live, sobald die Sprache in den Einstellungen
// umgeschaltet wird.

import Foundation

enum Loc {

    /// Aktuell gewählte Sprache ("de" / "en"). Wird von `AppState` gespiegelt,
    /// damit auch Klassen ohne Zugriff auf `AppState` (z. B. die Analyse-Engine)
    /// ihre sichtbaren Texte übersetzen können.
    static var aktuelleSprache: String = "de"

    /// Kurzform für Schichten ohne `AppState` (Analyse-Engine):
    /// liefert je nach aktueller Sprache den passenden Text.
    static func s(_ de: String, _ en: String) -> String {
        aktuelleSprache == "en" ? en : de
    }

    /// Übersetzungstabelle. Erster Wert = Deutsch, zweiter Wert = Englisch.
    static let tabelle: [String: (de: String, en: String)] = [

        // MARK: Seitenleiste (ContentView)
        "sidebar.section.analysis": ("Analyse", "Analysis"),
        "sidebar.analyse":          ("Datei prüfen", "Check file"),
        "sidebar.batch":            ("Stapelanalyse", "Batch analysis"),
        "sidebar.section.model":    ("Modell", "Model"),
        "sidebar.training":         ("Training", "Training"),
        "sidebar.visualization":    ("Visualisierung", "Visualization"),
        "sidebar.section.info":     ("Info", "Info"),
        "sidebar.settings":         ("Einstellungen", "Settings"),

        // MARK: Einstellungen (ContentView / SettingsView)
        "settings.title":        ("Einstellungen", "Settings"),
        "settings.autosave":     ("Änderungen werden automatisch gespeichert – kein Speichern-Button nötig.",
                                  "Changes are saved automatically – no save button needed."),
        "settings.appearance":   ("Darstellung", "Appearance"),
        "settings.orangeTheme":  ("Dunkles Orangethema aktiv", "Dark orange theme active"),
        "settings.preview":      ("Automatische Vorschau", "Automatic preview"),
        "settings.language":     ("Sprache", "Language"),
        "settings.german":       ("Deutsch", "German"),
        "settings.english":      ("Englisch", "English"),
        "settings.analysis":     ("Analyse", "Analysis"),
        "settings.batchConfirm": ("Batch-Bestätigung vor Start", "Confirm before batch start"),
        "settings.detailedHints":("Ausgabe detaillierter Hinweise", "Show detailed hints"),
        "settings.modeTrained":  ("Analysemodus: Trainiert", "Analysis mode: Trained"),

        // MARK: Menüleiste (TotallyHumanApp)
        "menu.analyzeFile":  ("Datei analysieren …", "Analyze file …"),
        "menu.openTraining": ("Training öffnen", "Open training"),
        "menu.quit":         ("Beenden", "Quit"),

        // MARK: Analyse-Ansicht
        "analysis.dropTitle":    ("Einzelanalyse", "Single analysis"),
        "analysis.dropSubtitle": ("Ziehe eine Audiodatei per Drag & Drop hierher oder klicke auf „Datei auswählen“.",
                                  "Drag & drop an audio file here or click “Select file”."),
        "analysis.title":        ("Analyse", "Analysis"),
        "analysis.subtitle":     ("Bewertet spektrale Artefakte und Hinweisindikatoren für KI-generierte Musik.",
                                  "Evaluates spectral artifacts and indicators for AI-generated music."),
        "verdict.ai":            ("KI-generiert", "AI-generated"),
        "verdict.real":          ("Echte Musik", "Real music"),
        "verdict.none":          ("Noch keine Analyse durchgeführt", "No analysis performed yet"),
        "label.aiMusic":         ("KI-Musik", "AI music"),
        "label.realMusic":       ("Echte Musik", "Real music"),
        "mode.trained":          ("Trainiert", "Trained"),
        "mode.heuristic":        ("Heuristisch", "Heuristic"),
        "metric.aiProbability":  ("KI-Wahrscheinlichkeit", "AI probability"),
        "metric.artifactScore":  ("Artefakt-Score", "Artifact score"),
        "metric.confidence":     ("Konfidenz", "Confidence"),
        "chart.spectrum":        ("Spektrum", "Spectrum"),
        "chart.frequency":       ("Frequenz", "Frequency"),
        "chart.level":           ("Pegel", "Level"),
        "chart.frequencyHz":     ("Frequenz (Hz)", "Frequency (Hz)"),
        "hints.title":           ("Hinweise", "Hints"),
        "hints.none":            ("Keine Hinweise.", "No hints."),
        "button.selectFile":     ("Datei auswählen", "Select file"),
        "result.details":        ("Details", "Details"),
        "result.file":           ("Datei", "File"),
        "result.classification": ("Einstufung", "Classification"),
        "result.mode":           ("Modus", "Mode"),
        "result.date":           ("Datum", "Date"),
        "result.none":           ("Kein Ergebnis vorhanden.", "No result available."),

        // MARK: Training-Ansicht
        "training.title":       ("Training", "Training"),
        "training.subtitle":    ("Füge echte und KI-generierte Musik hinzu und starte einen neuen Trainingslauf. Die Daten werden lokal gespeichert und wachsen mit jedem Durchlauf.",
                                 "Add real and AI-generated music and start a new training run. The data is stored locally and grows with each run."),
        "training.totalSamples":("Beispiele gesamt", "Total samples"),
        "training.realMusic":   ("Echte Musik", "Real music"),
        "training.aiMusic":     ("KI-Musik", "AI music"),
        "training.accuracy":    ("Genauigkeit", "Accuracy"),
        "training.modeTrained": ("Modus: Trainiert", "Mode: Trained"),
        "training.modeHeuristic":("Modus: heuristisch", "Mode: heuristic"),
        "training.folder":      ("Ordner", "Folder"),
        "training.folderHelp":  ("Öffnet den lokalen Ordner mit Modell und Trainingsdaten. Diese Dateien bleiben zwischen Programmstarts erhalten, sodass der Algorithmus dauerhaft dazulernt.",
                                 "Opens the local folder with the model and training data. These files persist between launches so the algorithm keeps learning."),
        "training.saveFailed":  ("Speichern fehlgeschlagen – Daten nur im Arbeitsspeicher.",
                                 "Saving failed – data only in memory."),
        "training.modelLocation":("Modell & Speicherort", "Model & storage location"),
        "training.dragOrClick": ("Ziehen oder klicken", "Drag or click"),
        "training.step1":       ("1 · Daten hinzufügen", "1 · Add data"),
        "training.needBoth":    ("Mindestens 1 echtes UND 1 KI-Beispiel benötigt.",
                                 "At least 1 real AND 1 AI sample required."),
        "training.start":       ("Training starten", "Start training"),
        "training.delete":      ("Löschen", "Delete"),
        "training.reseed":      ("Seed neu importieren", "Re-import seed"),
        "training.step2":       ("2 · Trainieren", "2 · Train"),
        "training.progress":    ("Verlauf", "Progress"),
        "training.run":         ("Lauf", "Run"),
        "training.seed":        ("Seed", "Seed"),
        "training.labelReal":   ("Echt", "Real"),
        "training.labelAI":     ("KI", "AI"),
        "training.addRealMusic":("Echte Musik +", "Real music +"),
        "training.addAIMusic":  ("KI-Musik +", "AI music +"),

        // MARK: Stapelanalyse
        "batch.title":       ("Stapelanalyse", "Batch analysis"),
        "batch.subtitle":    ("Analysiere mehrere Dateien in einem Durchgang und vergleiche die Ergebnisse direkt nebeneinander.",
                              "Analyze multiple files in one run and compare the results side by side."),
        "batch.dropTitle":   ("Ordner oder mehrere Dateien ablegen", "Drop a folder or multiple files"),
        "batch.dropSubtitle":("Unterstützt Drag & Drop für ganze Ordner und mehrere Audiodateien — oder klicke zum Auswählen.",
                              "Supports drag & drop for whole folders and multiple audio files — or click to select."),
        "batch.queue":       ("Warteschlange", "Queue"),
        "batch.queueEmpty":  ("Keine Dateien in der Warteschlange.", "No files in the queue."),
        "batch.ready":       ("Bereit", "Ready"),
        "batch.results":     ("Ergebnisse", "Results"),
        "batch.resultsEmpty":("Noch keine Batch-Ergebnisse vorhanden.", "No batch results yet."),
        "batch.ai":          ("KI", "AI"),
        "batch.distribution":("Verteilung", "Distribution"),

        // MARK: Visualisierung
        "viz.title":         ("Visualisierung", "Visualization"),
        "viz.subtitle":      ("Mehrere Diagramme für Spektrum, Zeitverlauf und Modellverhalten in einer Oberfläche.",
                              "Multiple charts for spectrum, time course and model behavior in one view."),
        "viz.classification":("Klassifikationsverlauf", "Classification course"),
        "viz.index":         ("Index", "Index"),
        "viz.value":         ("Wert", "Value"),
        "viz.segments":      ("Zeitliche Segmente", "Time segments"),
        "viz.segment":       ("Segment", "Segment"),
        "viz.strength":      ("Stärke", "Strength"),
        "viz.ssm":           ("SSM-Matrix", "SSM matrix"),
        "viz.noMatrix":      ("Keine Matrixdaten vorhanden.", "No matrix data available."),
    ]

    /// Liefert die Übersetzung für einen Schlüssel in der angegebenen Sprache.
    /// Fällt auf Deutsch bzw. den Schlüssel selbst zurück, falls unbekannt.
    static func t(_ key: String, _ sprache: String) -> String {
        guard let eintrag = tabelle[key] else { return key }
        return sprache == "en" ? eintrag.en : eintrag.de
    }
}

extension AppState {
    /// Komfort-Zugriff aus den Views: `appState.t("schlüssel")`.
    /// Nutzt automatisch die aktuell gewählte Sprache.
    func t(_ key: String) -> String {
        Loc.t(key, selectedLanguage)
    }

    /// Kurzform für dynamische Statustexte mit Interpolation:
    /// `s("Deutsch", "English")` liefert je nach aktueller Sprache den passenden Text.
    func s(_ de: String, _ en: String) -> String {
        selectedLanguage == "en" ? en : de
    }
}

File: Windows Source/TotallyHuman/Utilities/Localization.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace TotallyHuman.Utilities;

/// <summary>
/// Einfaches, Dictionary-basiertes Zweisprachigkeitssystem (Deutsch / Englisch).
/// Aufruf: <c>Loc.Get("key")</c>. Sprachwechsel via <c>Loc.SetLanguage("en")</c>
/// löst <see cref="LanguageChanged"/> aus, damit die UI sich neu bindet.
/// </summary>
public static class Loc
{
    public static string Language { get; private set; } = "de";

    public static event EventHandler? LanguageChanged;

    public static void SetLanguage(string lang)
    {
        lang = (lang == "en") ? "en" : "de";
        if (lang == Language) return;
        Language = lang;
        LanguageChanged?.Invoke(null, EventArgs.Empty);
    }

    public static string Get(string key)
    {
        var table = Language == "en" ? En : De;
        if (table.TryGetValue(key, out var v)) return v;
        // Fallback auf Deutsch, dann auf den Schlüssel selbst.
        return De.TryGetValue(key, out var d) ? d : key;
    }

    // Kurzform
    public static string T(string key) => Get(key);

    private static readonly Dictionary<string, string> De = new()
    {
        // App / Navigation
        ["app.title"] = "Totally Human — KI-Musik-Detektor",
        ["nav.section.analysis"] = "Analyse",
        ["nav.section.model"] = "Modell",
        ["nav.section.info"] = "Info",
        ["nav.analyse"] = "Datei prüfen",
        ["nav.batch"] = "Stapelanalyse",
        ["nav.training"] = "Training",
        ["nav.visualisierung"] = "Visualisierung",
        ["nav.einstellungen"] = "Einstellungen",
        ["status.ready"] = "Bereit",

        // Analyse-Tab
        ["analysis.title"] = "Analyse",
        ["analysis.subtitle"] = "Bewertet spektrale Artefakte und Hinweisindikatoren für KI-generierte Musik.",
        ["analysis.drop.title"] = "Einzelanalyse",
        ["analysis.drop.subtitle"] = "Ziehe eine Audiodatei per Drag & Drop hierher oder klicke auf „Datei auswählen“.",
        ["analysis.openfile"] = "Datei auswählen",
        ["analysis.noresult"] = "Noch keine Analyse durchgeführt",
        ["analysis.ki"] = "KI-generiert",
        ["analysis.human"] = "Echte Musik",
        ["analysis.kimusic"] = "KI-Musik",
        ["analysis.humanmusic"] = "Echte Musik",
        ["analysis.mode.trained"] = "Trainiert",
        ["analysis.mode.heuristic"] = "Heuristisch",
        ["analysis.metric.probability"] = "KI-Wahrscheinlichkeit",
        ["analysis.metric.artefact"] = "Artefakt-Score",
        ["analysis.metric.confidence"] = "Konfidenz",
        ["analysis.spectrum"] = "Spektrum",
        ["analysis.hints"] = "Hinweise",
        ["analysis.nohints"] = "Keine Hinweise.",
        ["analysis.details"] = "Details",
        ["analysis.noresult.details"] = "Kein Ergebnis vorhanden.",
        ["analysis.file"] = "Datei",
        ["analysis.classification"] = "Einstufung",
        ["analysis.mode"] = "Modus",
        ["analysis.date"] = "Datum",
        ["axis.freq"] = "Frequenz (Hz)",
        ["axis.db"] = "dB",

        // Batch-Tab
        ["batch.title"] = "Stapelanalyse",
        ["batch.subtitle"] = "Analysiere mehrere Dateien oder ganze Ordner gleichzeitig.",
        ["batch.choose"] = "Dateien / Ordner auswählen",
        ["batch.clear"] = "Liste leeren",
        ["batch.summary.total"] = "Dateien gesamt",
        ["batch.summary.ki"] = "KI-generiert",
        ["batch.summary.human"] = "Menschlich",
        ["batch.col.file"] = "Datei",
        ["batch.col.result"] = "Ergebnis",
        ["batch.col.probability"] = "KI-Wahrsch.",
        ["batch.col.confidence"] = "Konfidenz",
        ["batch.col.mode"] = "Modus",
        ["batch.empty"] = "Noch keine Dateien analysiert.",

        // Training-Tab
        ["training.title"] = "Training",
        ["training.subtitle"] = "Füge echte und KI-generierte Musik hinzu und starte einen neuen Trainingslauf. Die Daten werden lokal gespeichert.",
        ["training.total"] = "Beispiele gesamt",
        ["training.real"] = "Echte Musik",
        ["training.ki"] = "KI-Musik",
        ["training.accuracy"] = "Genauigkeit",
        ["training.mode.trained"] = "Modus: Trainiert",
        ["training.mode.heuristic"] = "Modus: heuristisch",
        ["training.openfolder"] = "Ordner öffnen",
        ["training.add.real"] = "Echte Musik hinzufügen",
        ["training.add.ki"] = "KI-Musik hinzufügen",
        ["training.dropreal"] = "Echte Musik hierher ziehen",
        ["training.dropki"] = "KI-Musik hierher ziehen",
        ["training.start"] = "Modell trainieren",
        ["training.reseed"] = "Seed-Daten neu importieren",
        ["training.clear"] = "Alle Trainingsdaten löschen",
        ["training.samples"] = "Trainingsproben",
        ["training.col.name"] = "Name",
        ["training.col.label"] = "Label",
        ["training.col.added"] = "Hinzugefügt",
        ["training.persist.error"] = "Speichern fehlgeschlagen – Daten nur im Arbeitsspeicher.",

        // Visualisierung-Tab
        ["viz.title"] = "Visualisierung",
        ["viz.subtitle"] = "Spektrum, Fingerprint und Segment-Stärken der zuletzt analysierten Datei.",
        ["viz.spectrum"] = "Spektrum & Fingerprint",
        ["viz.spectrum.line"] = "Spektrum (dB)",
        ["viz.fingerprint.line"] = "Fingerprint",
        ["viz.baseline.line"] = "Grundlinie",
        ["viz.segments"] = "Segment-Stärken",
        ["viz.segments.axis"] = "Segment",
        ["viz.strength"] = "Stärke",
        ["viz.empty"] = "Noch keine Analyse — bitte zuerst eine Datei im Analyse-Tab prüfen.",

        // Einstellungen-Tab
        ["settings.title"] = "Einstellungen",
        ["settings.autosave"] = "Änderungen werden automatisch gespeichert – kein Speichern-Button nötig.",
        ["settings.appearance"] = "Darstellung",
        ["settings.orangetheme"] = "Dunkles Orangethema aktiv",
        ["settings.preview"] = "Automatische Vorschau",
        ["settings.language"] = "Sprache",
        ["settings.lang.de"] = "Deutsch",
        ["settings.lang.en"] = "Englisch",
        ["settings.analysis"] = "Analyse",
        ["settings.quality"] = "Qualitätsschwelle",
        ["settings.segment"] = "Segmentlänge (Sekunden)",
        ["settings.batchconfirm"] = "Batch-Bestätigung vor Start",
        ["settings.detailedhints"] = "Ausgabe detaillierter Hinweise",
        ["settings.storage"] = "Speicher & Daten",
        ["settings.openfolder"] = "Speicherordner öffnen",
        ["settings.reseed"] = "Seed-Daten neu importieren",
        ["settings.about"] = "Über",
        ["settings.about.text"] = "Totally Human erkennt KI-generierte Musik anhand spektraler Artefakte im Hochfrequenzband (5–16 kHz) und einem logistischen Modell.",

        // Allgemein
        ["common.yes"] = "Ja",
        ["common.no"] = "Nein",
        ["common.busy"] = "Bitte warten …",
    };

    private static readonly Dictionary<string, string> En = new()
    {
        ["app.title"] = "Totally Human — AI Music Detector",
        ["nav.section.analysis"] = "Analysis",
        ["nav.section.model"] = "Model",
        ["nav.section.info"] = "Info",
        ["nav.analyse"] = "Check File",
        ["nav.batch"] = "Batch Analysis",
        ["nav.training"] = "Training",
        ["nav.visualisierung"] = "Visualization",
        ["nav.einstellungen"] = "Settings",
        ["status.ready"] = "Ready",

        ["analysis.title"] = "Analysis",
        ["analysis.subtitle"] = "Evaluates spectral artefacts and indicator metrics for AI-generated music.",
        ["analysis.drop.title"] = "Single Analysis",
        ["analysis.drop.subtitle"] = "Drag & drop an audio file here or click “Choose File”.",
        ["analysis.openfile"] = "Choose File",
        ["analysis.noresult"] = "No analysis performed yet",
        ["analysis.ki"] = "AI-generated",
        ["analysis.human"] = "Real Music",
        ["analysis.kimusic"] = "AI Music",
        ["analysis.humanmusic"] = "Real Music",
        ["analysis.mode.trained"] = "Trained",
        ["analysis.mode.heuristic"] = "Heuristic",
        ["analysis.metric.probability"] = "AI Probability",
        ["analysis.metric.artefact"] = "Artefact Score",
        ["analysis.metric.confidence"] = "Confidence",
        ["analysis.spectrum"] = "Spectrum",
        ["analysis.hints"] = "Hints",
        ["analysis.nohints"] = "No hints.",
        ["analysis.details"] = "Details",
        ["analysis.noresult.details"] = "No result available.",
        ["analysis.file"] = "File",
        ["analysis.classification"] = "Classification",
        ["analysis.mode"] = "Mode",
        ["analysis.date"] = "Date",
        ["axis.freq"] = "Frequency (Hz)",
        ["axis.db"] = "dB",

        ["batch.title"] = "Batch Analysis",
        ["batch.subtitle"] = "Analyze multiple files or entire folders at once.",
        ["batch.choose"] = "Choose Files / Folder",
        ["batch.clear"] = "Clear List",
        ["batch.summary.total"] = "Total Files",
        ["batch.summary.ki"] = "AI-generated",
        ["batch.summary.human"] = "Human",
        ["batch.col.file"] = "File",
        ["batch.col.result"] = "Result",
        ["batch.col.probability"] = "AI Prob.",
        ["batch.col.confidence"] = "Confidence",
        ["batch.col.mode"] = "Mode",
        ["batch.empty"] = "No files analyzed yet.",

        ["training.title"] = "Training",
        ["training.subtitle"] = "Add real and AI-generated music and start a new training run. Data is stored locally.",
        ["training.total"] = "Total Samples",
        ["training.real"] = "Real Music",
        ["training.ki"] = "AI Music",
        ["training.accuracy"] = "Accuracy",
        ["training.mode.trained"] = "Mode: Trained",
        ["training.mode.heuristic"] = "Mode: Heuristic",
        ["training.openfolder"] = "Open Folder",
        ["training.add.real"] = "Add Real Music",
        ["training.add.ki"] = "Add AI Music",
        ["training.dropreal"] = "Drag real music here",
        ["training.dropki"] = "Drag AI music here",
        ["training.start"] = "Train Model",
        ["training.reseed"] = "Re-import Seed Data",
        ["training.clear"] = "Delete All Training Data",
        ["training.samples"] = "Training Samples",
        ["training.col.name"] = "Name",
        ["training.col.label"] = "Label",
        ["training.col.added"] = "Added",
        ["training.persist.error"] = "Saving failed – data only in memory.",

        ["viz.title"] = "Visualization",
        ["viz.subtitle"] = "Spectrum, fingerprint and segment strengths of the most recently analyzed file.",
        ["viz.spectrum"] = "Spectrum & Fingerprint",
        ["viz.spectrum.line"] = "Spectrum (dB)",
        ["viz.fingerprint.line"] = "Fingerprint",
        ["viz.baseline.line"] = "Baseline",
        ["viz.segments"] = "Segment Strengths",
        ["viz.segments.axis"] = "Segment",
        ["viz.strength"] = "Strength",
        ["viz.empty"] = "No analysis yet — please check a file in the Analysis tab first.",

        ["settings.title"] = "Settings",
        ["settings.autosave"] = "Changes are saved automatically – no save button needed.",
        ["settings.appearance"] = "Appearance",
        ["settings.orangetheme"] = "Dark orange theme enabled",
        ["settings.preview"] = "Automatic preview",
        ["settings.language"] = "Language",
        ["settings.lang.de"] = "German",
        ["settings.lang.en"] = "English",
        ["settings.analysis"] = "Analysis",
        ["settings.quality"] = "Quality Threshold",
        ["settings.segment"] = "Segment Length (seconds)",
        ["settings.batchconfirm"] = "Confirm before batch start",
        ["settings.detailedhints"] = "Output detailed hints",
        ["settings.storage"] = "Storage & Data",
        ["settings.openfolder"] = "Open Storage Folder",
        ["settings.reseed"] = "Re-import Seed Data",
        ["settings.about"] = "About",
        ["settings.about.text"] = "Totally Human detects AI-generated music via spectral artefacts in the high-frequency band (5–16 kHz) and a logistic model.",

        ["common.yes"] = "Yes",
        ["common.no"] = "No",
        ["common.busy"] = "Please wait …",
    };
}
